toplevel/native/opttoploop.ml

```diff
@@ -307,13 +312,33 @@ let load_lambda ppf ~compilation_unit ~required_globals lam size =
   | None -> default_load ppf program
   | Some {Jit.load; _} -> load ppf program
 
+(* Don't call [outval_of_value] on non-values *)
+let outval_of_sig_value env id val_kind val_type =
+  let sort =
+    match val_kind with
+    | Val_reg sort -> sort
+    | Val_ivar _ -> Jkind_types.Sort.(of_const Const.for_instance_var)
+    | Val_self _ | Val_anc _ -> Jkind_types.Sort.(of_const Const.for_object)
+    | Val_prim _ ->
+      Misc.fatal_error "Unexpected primitive"
+    | Val_mut _ ->
+      Misc.fatal_error "Mutable variable found at the structure level"
+  in
+  let call_outval () = outval_of_value env (toplevel_value id) val_type in
+  match Jkind.Sort.default_for_transl_and_get sort with
+  | Base Value -> call_outval ()
+  | _ when not !Clflags.native_code -> call_outval()
+  | Base (Void | Untagged_immediate | Float64 | Float32 | Word | Bits8
+         | Bits16 | Bits32 | Bits64 | Vec128 | Vec256 | Vec512)
+  | Product _ -> Oval_stuff "<abstr>"
+
 (* Print the outcome of an evaluation *)
 
 let pr_item =
   Printtyp.print_items
     (fun env -> function
-       | Sig_value(id, {val_kind = Val_reg; val_type; _}, _) ->
-          Some (outval_of_value env (toplevel_value id) val_type)
+       | Sig_value(id, {val_kind = Val_reg sort; val_type; _}, _) ->
+         Some (outval_of_sig_value env id (Val_reg sort) val_type)
```

Comment:
As an extremely minor nit, this will save an allocation:
```suggestion
       | Sig_value(id, {val_kind = (Val_reg _) as vk; val_type; _}, _) ->
         Some (outval_of_sig_value env id vk val_type)
```